deepblast/viterbi.py

```diff
@@ -0,0 +1,163 @@
+import torch
+import torch.nn as nn
+from deepblast.ops import operators
+# from torch.nn.utils.rnn import pack_padded_sequence, PackedSequence
+# from deepblast.constants import x, m, y, s
+
+
+def _forward_pass(theta, A, pos, operator='softmax'):
+    """ Forward algorithm for K state HMM.
+
+    Parameters
+    ----------
+    theta : torch.Tensor
+        Input Potentials of dimension N x M x S. This represents the
+        pairwise residue distance across states.
+    A : torch.Tensor
+        Transition probabilities of dimensions N x M x S x S.
+        All of these parameters are assumed to be in log units
+    pos : list of int
+        Specifies the differential indices for each state.
+    operator : str
+        The smoothed maximum operator.
+    """
+    op = operators[operator]
+    new = theta.new
+    N, M, S = theta.size()
+    V = new(N + 1, M + 1, S).zero_()     # N x M x S
+    Q = new(N + 2, M + 2, S, S).zero_()  # N x M x S x S
+    Vt = new(S).zero_()    # S
+    neg_inf = -1e8   # very negative number
+    V = V + neg_inf
+    V[0, 0] = 0   # Make all states equally likely to enter
+    assert S == len(pos), f'{pos} does not have length {S}'
+    for i in range(1, N + 1):
+        for j in range(1, M + 1):
+            for k in range(S):
+                di, dj = pos[k]
+                res = op.max(V[i + di, j + dj] + A[i - 1, j - 1, k])
+                V[i, j, k], Q[i, j, k] = res
+            V[i, j] += theta[i - 1, j - 1]
+    # Terminate. First state is terminal state
+    Vt, Q[N + 1, M + 1, 0] = op.max(V[N, M])
+    return Vt, Q
```

Comment:
@blackwer to clarify what I mean, this function should return instead

```suggestion
    return Vt, Q, V
```

Turns out we need the full matrix in order to compute expectations for Baum-Welch.